plugins/single_plugins/scale.cpp

```diff
@@ -0,0 +1,1403 @@
+            {
+                next_focus = v;
+                break;
+            }
+        }
+
+        output->focus_view(next_focus, true);
+    }
+
+    bool animation_running()
+    {
+        for (auto& e : scale_data)
+        {
+            if (e.second.fade_animation.running() ||
+                e.second.animation.scale_animation.running())
+            {
+                return true;
+            }
+
+            for (auto& child : e.first->children)
+            {
+                if (scale_data[child].fade_animation.running() ||
+                    scale_data[child].animation.scale_animation.running())
+                {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    wf::effect_hook_t pre_hook = [=] ()
+    {
+        transform_views();
+    };
+
+    wf::effect_hook_t post_hook = [=] ()
+    {
+        output->render->schedule_redraw();
+
+        if (animation_running())
+        {
+            return;
+        }
+
+        unset_hook();
+
+        if (active)
+        {
+            return;
+        }
+
+        finalize();
+    };
+
+    bool activate()
+    {
+        if (active)
+        {
+            return false;
+        }
+
+        grab_interface->capabilities = wf::CAPABILITY_GRAB_INPUT;
+
+        if (!output->is_plugin_active(grab_interface->name) &&
+            !output->activate_plugin(grab_interface))
+        {
+            return false;
+        }
+
+        auto views = get_views();
+        if (!views.size())
+        {
+            output->deactivate_plugin(grab_interface);
+
+            return false;
+        }
+
+        initial_focus_view = output->get_active_view();
+        if (!interact)
+        {
+            if (!grab_interface->grab())
+            {
+                deactivate();
+
+                return false;
+            }
+
+            if (initial_focus_view)
+            {
+                output->focus_view(initial_focus_view, true);
+            }
+        }
+
+        active = true;
+
+        layout_slots(get_views());
+
+        if (interact)
+        {
+            connect_button_signal();
+        }
+
+        output->connect_signal("view-layer-attached", &view_attached);
+        output->connect_signal("view-attached", &view_attached);
+        view_detached.disconnect();
+        output->connect_signal("workspace-changed", &workspace_changed);
+        output->connect_signal("view-layer-detached", &view_detached);
+        output->connect_signal("view-minimized", &view_minimized);
+        output->connect_signal("view-unmapped", &view_unmapped);
+        output->connect_signal("view-focused", &view_focused);
+
+        view_geometry_changed.disconnect();
+        for (auto& e : scale_data)
+        {
+            auto view = e.first;
+            view->connect_signal("geometry-changed", &view_geometry_changed);
+            if ((view == initial_focus_view) || (view->parent == initial_focus_view))
```

Comment:
seeing this again here, maybe it is worth adding helper functions `get_topmost_view()` because there may be nested dialogues ...